### PR TITLE
[TorchToLinalg] Support lowering AtenMaxUnpool2dOp for linalg backend

### LINKED ISSUE #4255: [TorchToLinalg] Add support for AtenMaxUnpool2dOp and conversion to linalg dialect
Add support for AtenMaxUnpool2dOp and lowering to linalg backend

### PR DESCRIPTION
**Pull request by vinitdeodhar**
The change adds support to lower AtenMaxUnpool2dOp for linalg backend.  Existing lowering for AtenMaxUnpool3dOp is refactored such that 2d/3d variants share utils 

Unlike max_unpool3d, max_unpool2d does not support specification of stride, padding
[torch op registry](https://github.com/pytorch/pytorch/blob/7451dd058564b5398af79bfc1e2669d75f9ecfa2/torch/csrc/jit/passes/utils/op_registry.cpp#L21)

The default values of stride and padding
https://docs.pytorch.org/docs/stable/generated/torch.nn.MaxUnpool2d.html

Closes #4255

### CODE REVIEW

**Comment by vinitdeodhar** [comment]
@vivekkhandelwal1 , @zjgarvey , @sahas3 can you please provide a review ?

**Comment by vinitdeodhar** [comment]
@vivekkhandelwal1 , @zjgarvey , @sahas3 can you please review the PR ?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 1]
```diff
@@ -679,156 +846,74 @@ class ConvertAtenMaxUnpool3dOp final
...
-            .create<linalg::GenericOp>(loc,
-                                       /*resultTensorTypes=*/init.getType(),
-                                       /*inputs=*/ValueRange({self, indices}),
-                                       /*outputs=*/init,
-                                       /*indexingMaps=*/indexingMaps,
-                                       /*iteratorTypes=*/iteratorTypes, builder)
-            .getResult(0);
+    auto resType = typeConverter->convertType<RankedTensorType>(op.getType());
+    if (!resType)
+      return rewriter.notifyMatchFailure(op, "invalid result type");
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp]: Isn't this already validated in `verifyLinalgCompatibleTypes` ?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 2]
```diff
@@ -679,156 +846,74 @@ class ConvertAtenMaxUnpool3dOp final
...
-      AffineExpr dim = rewriter.getAffineDimExpr(i);
-      if (i < NC) {
-        inputExprs.emplace_back(dim);
-      } else {
-        int64_t j = i - NC;
-        inputExprs.emplace_back(dim.floorDiv(stride[j]));
-      }
-      outputExprs.emplace_back(dim);
-    }
+    ArrayRef<int64_t> inputSize =
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:875]: This is checking for spatial dimensions -- consider naming this as `spatialInputSize` for clarity.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 3]
```diff
@@ -2461,6 +2461,34 @@ def AdaptiveMaxPool3dStaticWithIndices_basic(module, tu: TestUtils):
...
+    @export
+    @annotate_args(
+        [
+            None,
+            ([2, 2, 2, 4], torch.float32, True),
+            ([2, 2, 2, 4], torch.int64, True),
+        ]
+    )
+    def forward(self, x, indices):
+        return torch.ops.aten.max_unpool2d(x, indices, (4, 8))
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2477]: Can you add a test for not providing output size?

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 4]
```diff
@@ -1101,6 +1101,15 @@ def aten〇max_unpool3d〡shape(self: List[int], indices: List[int], output_size
     else:
         return [self[0], output_size[0], output_size[1], output_size[2]]
 
+def aten〇max_unpool2d〡shape(self: List[int], indices: List[int], output_size: List[int]) -> List[int]:
+    assert (len(self) == 4 or len(self) == 3), "Input be of rank 3 or 4"
+    assert (len(output_size) == 2), "output_size must have 2 elements"
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1106]: Not much familiar with this part of the code -- When `output_size` is not provided when constructing the op in python, is this automatically set from python?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 5]
```diff
@@ -215,6 +215,177 @@ static LogicalResult createPoolingOp(
   return success();
 }
 
+static Value createMaxUnpoolOp(Operation *op, int64_t poolingDimensionality,
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Pooling.cpp:218]: This change looks just refactoring of existing code without any additional changes?

Can maxUnpool1d share this implementation as well? Not asking to implement it here, just for clarification.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py [hunk 6]
```diff
@@ -2461,6 +2461,34 @@ def AdaptiveMaxPool3dStaticWithIndices_basic(module, tu: TestUtils):
...
 
+class MaxUnpool2dModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args(
+        [
+            None,
+            ([2, 2, 2, 4], torch.float32, True),
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2472]: Is dynamic batch/channel supported? The lowering only checks for static spatial dims it seems.

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 7]
```diff
@@ -1101,6 +1101,15 @@ def aten〇max_unpool3d〡shape(self: List[int], indices: List[int], output_size
     else:
         return [self[0], output_size[0], output_size[1], output_size[2]]
 
+def aten〇max_unpool2d〡shape(self: List[int], indices: List[int], output_size: List[int]) -> List[int]:
+    assert (len(self) == 4 or len(self) == 3), "Input be of rank 3 or 4"
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1105]: Can you add an e2e test for rank 3?

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 1 above)
Comment (vinitdeodhar) [reply lib/Conversion/TorchToLinalg/Pooling.cpp]: Thanks for pointing it out, this is validated in verifyLinalgCompatibleTypes

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py (hunk 3 above)
Comment (vinitdeodhar) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2477]: output size is required parameter for ops.aten.max_unpool2d

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 5 above)
Comment (vinitdeodhar) [reply lib/Conversion/TorchToLinalg/Pooling.cpp:218]: Yes, this is refactoring of existing code without any additional changes. MaxUnpool1d is not supported currently, but a future implementation can share this implementation

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py (hunk 7 above)
Comment (vinitdeodhar) [reply projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1105]: Added a e2e test for rank 3

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py (hunk 4 above)
Comment (vinitdeodhar) [reply projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1106]: When output_size is not provided, it is inferred from python. The documentation has a note about the behavior: https://docs.pytorch.org/docs/stable/generated/torch.nn.MaxUnpool3d.html

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py (hunk 3 above)
Comment (sahas3) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2477]: Ah I was referring to the layer itself -- here is a doc example showing that it's an optional input https://docs.pytorch.org/docs/stable/generated/torch.nn.MaxUnpool2d.html#torch.nn.MaxUnpool2d ? As per your other comment, seems like it will be resolved to some value when the aten IR is produced but it'll be good to lock it down.

**Comment by vinitdeodhar** [comment]
@vivekkhandelwal1 , @zjgarvey can you please review ?

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 8]
```diff
@@ -215,6 +215,177 @@ static LogicalResult createPoolingOp(
...
+    }
+    int64_t pad = padding[i - NC];
+
+    outSizePadded.emplace_back(
+        rewriter.create<arith::ConstantIndexOp>(loc, size + pad));
+  }
+
+  auto ceilDiv = [](int64_t v1, int64_t v2) -> int64_t {
+    return (v1 + v2 - 1) / v2;
+  };
```
Comment (zjgarvey) [inline lib/Conversion/TorchToLinalg/Pooling.cpp]: Nit: It looks like this code was moved from elsewhere, but I'm not sure I understand the point of using this lambda when it only has one use, and the naming is not very appropriate. We could just inline the simple arithmetic below.

File: lib/Conversion/TorchToLinalg/Pooling.cpp [hunk 9]
```diff
@@ -613,19 +784,18 @@ class ConvertAtenMaxUnpool3dOp final
...
       return failure();
 
-    Location loc = op->getLoc();
     const TypeConverter *typeConverter = getTypeConverter();
     Value self = adaptor.getSelf();
     auto selfType = cast<RankedTensorType>(self.getType());
 
-    ArrayRef<int64_t> inputSize = selfType.getShape().take_back(3);
-    if (ShapedType::isDynamicShape(inputSize))
+    ArrayRef<int64_t> spatialInputSize = selfType.getShape().take_back(3);
```
Comment (zjgarvey) [inline lib/Conversion/TorchToLinalg/Pooling.cpp]: nit: If we are changing this name, maybe `spatialInputShape` would be more appropriate.

File: lib/Conversion/TorchToLinalg/Pooling.cpp (hunk 8 above)
Comment (vinitdeodhar) [reply lib/Conversion/TorchToLinalg/Pooling.cpp]: Yes, this is just used once. Inlined the lambda function at its callsite

**Comment by vinitdeodhar** [comment]
@zjgarvey , @sahas3 thanks for your feedback, can you help merge this change ?

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py (hunk 6 above)
Comment (sahas3) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2472]: @vinitdeodhar can you clarify this question? Otherwise your changes looks good to me. Thanks!

Comment (vinitdeodhar) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/pooling.py:2472]: @sahas3 , dynamic batch/channel is not supported and the lowering checks and errors out for the case. The limitation was inherited from the existing unpooling code reused by this lowering
